client/app/components/queries/visualization-embed.js

```diff
@@ -8,47 +8,46 @@ const VisualizationEmbed = {
   bindings: {
     data: '<',
   },
-  controller($routeParams, Query, QueryResult) {
+  controller($http, $q, $routeParams, Auth, Query, QueryResult) {
     'ngInject';
 
     document.querySelector('body').classList.add('headless');
     const visualizationId = parseInt($routeParams.visualizationId, 10);
     this.showQueryDescription = $routeParams.showDescription;
-    this.apiKey = $routeParams.api_key;
     this.logoUrl = logoUrl;
-    this.query = new Query(this.data[0]);
-    this.queryResult = new QueryResult(this.data[1]);
-    this.visualization =
-      find(this.query.visualizations, visualization => visualization.id === visualizationId);
+
+    this.apiKey = $routeParams.api_key;
+    Auth.setApiKey(this.apiKey);
+    Auth.loadConfig().then(() => {
+      const queryId = $routeParams.queryId;
+
+      const query = $http.get(`api/queries/${queryId}`).then(response => response.data);
+      const queryResult = $http.post(`api/queries/${queryId}/results`, {
+        parameters: queryStringParameters(),
+      }).then(response => response.data, (error) => {
+        if (error.status === 400) {
+          return {};
+        }
+
+        throw error;
```

Comment:
Previously, an unhandled promise rejection here would render a full-page error message.
Now, after the data-loading code has moved into the controller, the page layout is already loaded and re-throwing this error does not render an error layout. @kravets-levko / @ranbena could you please give me a little guidance on how to render a full-page error message here?

Reply:
@rauchy here is a good example how to do it: https://github.com/getredash/redash/blob/f9cc230227b42799886decdf315dea5c273773f3/client/app/pages/data-sources/EditDataSource.jsx
1. inject Angular's `$exceptionHandler` service;
2. if it's HTTP error - you need to convert it to `PromiseRejectionError`
3. pass the error to `$exceptionHandler`

It should be something like this:
```js
// ANGULAR_REMOVE_ME This code is related to Angular's HTTP services
if (error.status && error.data) {
  error = new PromiseRejectionError(error);
}
$exceptionHandler(error);
```